### PR TITLE
Add framework for preloading controllers

### PR DESCRIPTION
**Pull request by kraih**
### Summary
This is a less ambitious solution for #1607. It doesn't preload all possible controllers but leaves it up to the developer to configure the right namespaces. The default of `MyApp::Controller` might just work for most of our users though.

### Motivation
More efficient memory usage. Detecting syntax errors in controllers much earlier (during application startup instead of at runtime).

### References
#1607

### CODE REVIEW

**Comment by amenonsen** [comment]
For what it's worth, I read through this commit, and it looks good to me.

File: lib/Mojo/Loader.pm [hunk 1]
```diff
@@ -15,12 +15,17 @@ sub data_section { $_[0] ? $_[1] ? _all($_[0])->{$_[1]} : _all($_[0]) : undef }
 sub file_is_binary { keys %{_all($_[0])} ? !!$BIN{$_[0]}{$_[1]} : undef }
 
 sub find_modules {
-  my $ns = shift;
+  my ($ns, $options) = (shift, shift // {});
+
+  my @ns  = split '::', $ns;
```
Comment (Grinnz) [inline lib/Mojo/Loader.pm:20]: `split /::/, $ns;`

**Comment by Grinnz** [comment]
I like this approach especially as it can be easily configured to avoid preloading if someone was so inclined (e.g. CGI deployment).

**Review by christopherraa — APPROVED** [review]
:+1: 

No objections from me. Surprised that this new functionality wasn't a bigger change.

**Comment by christopherraa** [comment]
One question as an after-thought: why default to `MyApp::Controller`? Does anyone really use that? If not, would it make more sense to default to whatever namespace the app is in, in the case of a full app? So if anyone has generated a `Corp::App::Foo` application, defaulting to preloading `Corp::App::Foo`. Too difficult or error-prone to detect perhaps?

File: lib/Mojo/Loader.pm (hunk 1 above)
Comment (kraih) [reply lib/Mojo/Loader.pm:20]: We actually use the string form everywhere in Mojolicious.

**Comment by kraih** [comment]
When i say `MyApp::Controller` i of course mean `"@{[ref $self]}::Controller"` in code.

File: lib/Mojo/Loader.pm (hunk 1 above)
Comment (Grinnz) [reply lib/Mojo/Loader.pm:20]: Weird, I'll suggest changing that separately...

**Review by Grinnz — APPROVED** [review]

**Review by marcusramberg — COMMENTED** [review]
I think this is a great addition to mojo. 👍

**Pull request merged**